From 8a8b9608f0311f44950a22994e75a2c5528bb7a1 Mon Sep 17 00:00:00 2001
From: Vladimir Lypak <vladimir.lypak@gmail.com>
Date: Sat, 23 Aug 2025 09:46:28 +0000
Subject: [PATCH 10/20] MSM8953 drm/msm/dsi: Reset DSI host on video done
 timeout

If video mode is used, DSI host needs a reset after encoder shut down
to be able to transmit commands again. If it's not done we will time
out waiting on VIDEO_DONE IRQ because VIDEO_MODE_ENGINE_BUSY flag gets
stuck after encoder is disabled. Subsequent command transfers time out
too:
[   89.798035] msm_dsi 1a94000.dsi: [drm:dsi_cmds2buf_tx [msm]] *ERROR* wait for video done timed out
[   90.005916] dsi_cmds2buf_tx: cmd dma tx failed, type=0x15, data0=0x28, len=4, ret=-110

To work around this issue we can reset DSI host if VIDEO_DONE IRQ didn't
trigger after at least one refresh period has passed. This fixes command
transfers from panel unprepare callback.

Also since we legitimizing the time out let's reduce delay in reset.
In downstream kernel the reset delay is at most 1 microsecond.

Fixes: 007ac0262b0d ("drm/msm/dsi: switch to DRM_PANEL_BRIDGE")
Signed-off-by: Vladimir Lypak <vladimir.lypak@gmail.com>
---
 drivers/gpu/drm/msm/dsi/dsi_host.c | 25 ++++++++++++++++---------
 1 file changed, 16 insertions(+), 9 deletions(-)

diff --git a/drivers/gpu/drm/msm/dsi/dsi_host.c b/drivers/gpu/drm/msm/dsi/dsi_host.c
index 89face2..c1b7070 100644
--- a/drivers/gpu/drm/msm/dsi/dsi_host.c
+++ b/drivers/gpu/drm/msm/dsi/dsi_host.c
@@ -32,8 +32,6 @@
 #include "msm_gem.h"
 #include "phy/dsi_phy.h"
 
-#define DSI_RESET_TOGGLE_DELAY_MS 20
-
 static int dsi_populate_dsc_params(struct msm_dsi_host *msm_host, struct drm_dsc_config *dsc);
 
 static int dsi_get_version(const void __iomem *base, u32 *major, u32 *minor)
@@ -166,6 +164,7 @@ struct msm_dsi_host {
 
 	struct drm_display_mode *mode;
 	struct drm_dsc_config *dsc;
+	unsigned long vid_done_timeout;
 
 	/* connected device info */
 	unsigned int channel;
@@ -1158,7 +1157,7 @@ static void dsi_sw_reset(struct msm_dsi_host *msm_host)
 
 	/* dsi controller can only be reset while clocks are running */
 	dsi_write(msm_host, REG_DSI_RESET, 1);
-	msleep(DSI_RESET_TOGGLE_DELAY_MS); /* make sure reset happen */
+	udelay(5); /* make sure reset happen */
 	dsi_write(msm_host, REG_DSI_RESET, 0);
 	wmb(); /* controller out of reset */
 
@@ -1212,17 +1211,19 @@ static void dsi_wait4video_done(struct msm_dsi_host *msm_host)
 	u32 ret = 0;
 	struct device *dev = &msm_host->pdev->dev;
 
-	dsi_intr_ctrl(msm_host, DSI_IRQ_MASK_VIDEO_DONE, 1);
-
 	reinit_completion(&msm_host->video_comp);
 
-	ret = wait_for_completion_timeout(&msm_host->video_comp,
-			msecs_to_jiffies(70));
+	dsi_intr_ctrl(msm_host, DSI_IRQ_MASK_VIDEO_DONE, 1);
 
-	if (ret == 0)
-		DRM_DEV_ERROR(dev, "wait for video done timed out\n");
+	ret = wait_for_completion_timeout(&msm_host->video_comp,
+			msm_host->vid_done_timeout);
 
 	dsi_intr_ctrl(msm_host, DSI_IRQ_MASK_VIDEO_DONE, 0);
+
+	if (ret == 0) {
+		DRM_DEV_DEBUG(dev, "wait for video done timed out, resetting dsi\n");
+		dsi_sw_reset(msm_host);
+	}
 }
 
 static void dsi_wait4video_eng_busy(struct msm_dsi_host *msm_host)
@@ -2018,6 +2019,7 @@ int msm_dsi_host_init(struct msm_dsi *msm_dsi)
 	if (!msm_host)
 		return -ENOMEM;
 
+	msm_host->vid_done_timeout = msecs_to_jiffies(50);
 	msm_host->pdev = pdev;
 	msm_dsi->host = &msm_host->base;
 
@@ -2589,6 +2591,7 @@ int msm_dsi_host_set_display_mode(struct mipi_dsi_host *host,
 				  const struct drm_display_mode *mode)
 {
 	struct msm_dsi_host *msm_host = to_msm_dsi_host(host);
+	u64 refresh_period = USEC_PER_SEC;
 
 	if (msm_host->mode) {
 		drm_mode_destroy(msm_host->dev, msm_host->mode);
@@ -2601,6 +2604,10 @@ int msm_dsi_host_set_display_mode(struct mipi_dsi_host *host,
 		return -ENOMEM;
 	}
 
+	refresh_period = mult_frac(refresh_period, (u64) mode->clock,
+			mode->htotal * (u64) mode->vtotal);
+	msm_host->vid_done_timeout = usecs_to_jiffies(refresh_period);
+
 	return 0;
 }
 
-- 
2.56.0

